sql: node crash with explain

The node serving the request crashed if I try to explain a query on beta-20170309
```sql
CREATE TABLE Test (ID serial PRIMARY KEY, Date TimeStampTZ,INDEX date_idx (Date));
insert into Test (Date) values ('2017-03-17 15:33:00'::timestamptz);

explain select * from Test where Date >='2017-01-01 00:00:00'::TimeStamp;
```

The node will crash and in the log, I see
```
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range

goroutine 26996 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4203634a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0x6e
panic(0x1ab98e0, 0x2878ec0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc42209c200, 0x7f614d43acc0, 0xc4398842a0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:340 +0x74
panic(0x1ab98e0, 0x2878ec0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).sendResponse(0xc42209c200, 0x7f614d43acc0, 0xc4398842a0, 0xc421790aa0, 0x1, 0x1, 0x2dd4cd8, 0x0, 0x0, 0xc4348a3600, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:961 +0x497
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc42209c200, 0x7f614d43acc0, 0xc4398842a0, 0xc4348a364d, 0x48, 0xc42683b518, 0x2dd4cd8, 0x0, 0x0, 0x1acd900, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:828 +0x202
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleExecute(0xc42209c200, 0x7f614d43acc0, 0xc4398842a0, 0xc42209c228, 0x9, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:805 +0x19d
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc42209c200, 0x7f614d43acc0, 0xc4398842a0, 0xc429460620, 0x5400, 0xc4202e0160, 0x7f614d43ada0, 0xc42942d300, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:442 +0xa7b
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4202e0000, 0x7f614d43ada0, 0xc42942d300, 0x28bed80, 0xc4261a2140, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:415 +0x917
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x28bed80, 0xc4261a2140)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:544 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4203634a0, 0xc42000e040, 0x28bed80, 0xc4261a2140, 0xc421194240)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x239
```

--- a/pkg/sql/pgwire/pgwire_test.go
+++ b/pkg/sql/pgwire/pgwire_test.go
@@ -731,6 +731,10 @@ func TestPGPreparedQuery(t *testing.T) {
 		"SELECT * FROM d.empty": {
 			baseTest.SetArgs(),
 		},
+		// #14238
+		"EXPLAIN SELECT 1": {
+			baseTest.SetArgs().Results(0, "render", "", "").Results(1, "nullrow", "", ""),
+		},
 
 		// TODO(jordan) blocked on #13651
 		//"SELECT $1::INT[]": {

--- a/pkg/sql/pgwire/v3.go
+++ b/pkg/sql/pgwire/v3.go
@@ -786,6 +786,11 @@ func (c *v3Conn) handleBind(ctx context.Context, buf *readBuffer) error {
 	if err != nil {
 		return err
 	}
+
+	if log.V(2) {
+		log.Infof(ctx, "portal: %q for %q, args %q, formats %q", portalName, stmt.Query, qargs, columnFormatCodes)
+	}
+
 	// Attach pgwire-specific metadata to the PreparedPortal.
 	portal.ProtocolMeta = preparedPortalMeta{outFormats: columnFormatCodes}
 	c.writeBuf.initMsg(serverMsgBindComplete)

--- a/pkg/sql/plan.go
+++ b/pkg/sql/plan.go
@@ -365,6 +365,10 @@ func (p *planner) newPlan(
 	}
 }
 
+// prepare constructs the logical plan for the statement.  This is
+// needed both to type placeholders and to inform pgwire of the types
+// of the result columns. All statements that either support
+// placeholders or have result columns must be handled here.
 func (p *planner) prepare(ctx context.Context, stmt parser.Statement) (planNode, error) {
 	if plan, err := p.maybePlanHook(ctx, stmt); plan != nil || err != nil {
 		return plan, err
@@ -373,6 +377,8 @@ func (p *planner) prepare(ctx context.Context, stmt parser.Statement) (planNode,
 	switch n := stmt.(type) {
 	case *parser.Delete:
 		return p.Delete(ctx, n, nil, false)
+	case *parser.Explain:
+		return p.Explain(ctx, n, false)
 	case *parser.Help:
 		return p.Help(ctx, n)
 	case *parser.Insert:
@@ -406,8 +412,9 @@ func (p *planner) prepare(ctx context.Context, stmt parser.Statement) (planNode,
 	case *parser.Update:
 		return p.Update(ctx, n, nil, false)
 	default:
-		// Other statement types do not support placeholders so there is no need
-		// for any special handling here.
+		// Other statement types do not have result columns and do not
+		// support placeholders so there is no need for any special
+		// handling here.
 		return nil, nil
 	}
 }

